### PR TITLE
Add a maistra istio 2.4 presubmit integ job

### PR DESCRIPTION
**Pull request by yxun**
Signed-off-by: Yuanlin <yuanlin.xu@redhat.com>

### CODE REVIEW

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-istio-integ-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-3

File: ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml [hunk 1]
```diff
@@ -1,18 +1,28 @@
+base_images:
+  cli:
+    name: "4.10"
```
Comment (jwendell) [inline ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml:3]: Does it make sense use ocp 4.11 as it's the latest?

File: ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml [hunk 2]
```diff
@@ -49,6 +60,74 @@ tests:
     ISTIO_ENVOY_BASE_URL="https://storage.googleapis.com/maistra-prow-testing/proxy"
   container:
     from: src
+  timeout: 20m0s
+- as: maistra-istio-integ-aws-2-3
```
Comment (jwendell) [inline ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml:64]: IMO we should start these tests in 2.4.

File: ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml [hunk 3]
```diff
@@ -49,6 +60,74 @@ tests:
...
+      from: src
+      resources:
+        requests:
+          cpu: 100m
+          memory: 100Mi
+      timeout: 10m0s
+    - as: maistra-istio-integration-kube-sc-2-3
+      cli: latest
+      commands: |
+        MAISTRA_NAMESPACE="maistra-e2e-test"
```
Comment (jwendell) [inline ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml:97]: ditto

File: ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml [hunk 4]
```diff
@@ -49,6 +60,74 @@ tests:
...
+      from: src
+      resources:
+        requests:
+          cpu: "2"
+          memory: 1Gi
+      timeout: 2h0m0s
+    - as: maistra-istio-integration-servicemesh-kube-2-3
+      cli: latest
+      commands: |
+        MAISTRA_MC_POD="maistra-e2e-test-mc-pod"
```
Comment (jwendell) [inline ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml:115]: ditto

File: ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-commands.sh [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        if [[ $((CNT)) -eq 0 ]]; then
+            echo "Pod $1 did not successfully deploy"
+            oc -n "${MAISTRA_NAMESPACE}" get pods "$1"
+            return 1
+        fi
+    done
+}
+
+
+MAISTRA_NAMESPACE=$(
```
Comment (jwendell) [inline ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-commands.sh:42]: Why redefining this variable to its same value?

File: ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-ref.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  env:
+  - name: MAISTRA_BUILDER_IMAGE # the value is defined in the workflow env
+  - name: MAISTRA_NAMESPACE
+    default: "maistra-e2e-test"
+  - name: MAISTRA_SC_POD
+    default: "maistra-e2e-test-sc-pod"
+  - name: MAISTRA_MC_POD
+    default: "maistra-e2e-test-mc-pod"
+  - name: GOFLAGS
+    default: '-mod=vendor'
```
Comment (jwendell) [inline ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-ref.yaml:18]: In 2.4 we don't need this variable

File: ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-ref.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  - name: BUILD_WITH_CONTAINER
+    default: '0'
+  - name: INTEGRATION_TEST_FLAGS
+    default: ' --istio.test.retries=1 '
+  - name: ISTIO_ENVOY_BASE_URL
```
Comment (jwendell) [inline ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-ref.yaml:23]: In 2.4 we don't use this variable yet

File: ci-operator/step-registry/servicemesh/istio/post/servicemesh-istio-post-commands.sh [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+set -o nounset
+set -o errexit
+set -o pipefail
+
+MAISTRA_NAMESPACE=$(oc delete project ${MAISTRA_NAMESPACE})
```
Comment (jwendell) [inline ci-operator/step-registry/servicemesh/istio/post/servicemesh-istio-post-commands.sh:7]: I still don't understand what's the purpose of reassign the variable.

File: ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml [hunk 9]
```diff
@@ -49,6 +60,74 @@ tests:
...
+    env:
+      MAISTRA_BUILDER_IMAGE: registry.ci.openshift.org/ci/maistra-builder:2.3
+      MAISTRA_MC_POD: maistra-e2e-test-mc-pod
+      MAISTRA_NAMESPACE: maistra-e2e-test
+      MAISTRA_SC_POD: maistra-e2e-test-sc-pod
+    test:
+    - as: maistra-istio-integration-copy-src
+      cli: latest
+      commands: |
+        MAISTRA_MC_POD="maistra-e2e-test-mc-pod"
```
[the remark this answers is not in the document]
Comment (yxun) [reply ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml:82]: yes, the env definition above is visible only for a workflow level. But in the tests level, I still got unbound var error without repeating them there.

File: ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-commands.sh (hunk 5 above)
Comment (yxun) [reply ci-operator/step-registry/servicemesh/istio/e2e/servicemesh-istio-e2e-commands.sh:42]: Just for waiting the oc create command finish its ns creation. I can remove that way and put it into a function.

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.3-maistra-istio-integ-aws-2-4

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-integ-aws-2-4

File: ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml (hunk 1 above)
Comment (yxun) [reply ci-operator/config/maistra/istio/maistra-istio-maistra-2.3.yaml]: From https://docs.ci.openshift.org/docs/how-tos/cluster-claim/#existing-cluster-pools
It looks like there are more standby or ready OCP 4.10 clusters than OCP 4.11 ones.
That means if we use ocp 4.11 we may need to wait more time in line when no standby cluster in the pool.

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-integ-aws-2-4

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-integ-aws-2-4

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-integ-aws-2-4

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-integ-aws-2-4

**Review by jwendell — APPROVED** [review]

**Review by droslean — APPROVED** [review]

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-lint-2-4

**Comment by yxun** [comment]
/pj-rehearse pull-ci-maistra-istio-maistra-2.4-maistra-istio-integ-aws-2-4

**Pull request merged**